src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs

```diff
@@ -101,66 +113,112 @@ private bool RunWorkInParallel()
         if (_runWorkload == null)
             throw new InvalidOperationException($"{nameof(_runWorkload)} was not provided.");
 
-        // Reserve slots and assign them work under the lock so we keep
-        // the slots consistent.
-        List<SlotWorkloadPair> workToRun = new();
+        // Reserve slots and assign them work under the lock so we keep the slots consistent.
+        List<Slot> slots;
         lock (_lock)
         {
-            if (_workloads.Count == 0)
-                return false;
-
             // When HandlePartialDiscovery or HandlePartialRun are in progress, and we call StopAllManagers,
             // it is possible that we will clear all slots, and have RunWorkInParallel waiting on the lock,
             // so when it is allowed to enter it will try to add more work, but we already cancelled,
             // so we should not start more work.
             if (!_acceptMoreWork)
                 return false;
 
-            var availableSlots = _managerSlots.Where(slot => slot.IsAvailable).ToList();
+            // We grab all empty slots.
+            var availableSlots = _managerSlots.Where(slot => !slot.HasWork).ToList();
```

Comment:
I would have used `ToArray` or `ToImmutableArray` as the object is lighter and it's showing better your intent of only reading length and some items.

Reply:
This is true for the various calls below. Feel free to ignore if you are not changing anything else.

Reply:
Fixed, thanks !